builtin/common/vector.lua

```diff
@@ -141,3 +141,72 @@ function vector.sort(a, b)
 	return {x = math.min(a.x, b.x), y = math.min(a.y, b.y), z = math.min(a.z, b.z)},
 		{x = math.max(a.x, b.x), y = math.max(a.y, b.y), z = math.max(a.z, b.z)}
 end
+
+function vector.rotate_around_axis(v, axis, angle)
+	--flip the angle because the formula is for the right hand rule
+	angle = -angle
+	local cosangle = math.cos(angle)
+	local sinangle = math.sin(angle)
+	axis = vector.normalize(axis)
+	--https://en.wikipedia.org/wiki/Rodrigues%27_rotation_formula
+	return vector.add(vector.add(
+		vector.multiply(v, cosangle),
+		vector.multiply(vector.cross(axis, v), sinangle)),
+		vector.multiply(axis, vector.dot(axis, v) * (1 - cosangle)))
```

Comment:
`a * cosangle + b * (1 - cosangle)` can usually be shortened to
`cosangle * (a - b) + b` where `a` and `b` are vectors calculated by `vector.multiply` above.

Before: (ignoring `vector.dot`)
 - 7 multiplications (6 in vector, 1 simple)
 - 1 subtraction (simple)
 - 3 additions (vector)

After: (ignoring `vector.dot`)
 - 3 subtractions (vector)
 - 3 multiplications (vector)
 - 3 additions (vector)

```Lua
local dot_axis = vector.multiply(axis, vector.dot(axis, v))
return vector.add(
	vector.multiply(vector.cross(axis, v), sinangle)),
	vector.add(
		vector.multiply(cosangle, vector.subtract(v, dot_axis))
		dot_axis
	)
```
PS: Not tested, but shouldn't this work?

Reply:
You could maybe test this by choosing vectors with only a small angle between axis and vector. Then compare if the angle between before and after. If it doesn't stay the same, this might be less precise.
```lua
local axis = vector.new(1, 0, 0)
local v = vector.new(420, 1, 0)
local rotated = vector.rotate_around_axis(v, axis, 123)
local rotated_two_times = vector.rotate_around_axis(rotated, axis, 345)
local old_angle = vector.angle(axis, v)
assert.equal(old_angle, vector.angle(axis, rotated))
assert.equal(old_angle, vector.angle(axis, rotated_two_times))
```